sip/sip-002-smart-contract-language.md

```diff
@@ -0,0 +1,309 @@
+transactions would be contentious and potentially create a rift in the
+community, especially as it will not be easy to deduce which contracts
+exactly were affected and for how long. In contrast, bugs in the VM
+itself present a more clear case for a hard fork: the smart contract
+was defined correctly, as everyone can see directly on the chain, but
+illegal transactions were incorrectly marked as valid.
+
+# Example: Simple Naming System
+
+To demonstrate the expressiveness of this smart contracting language,
+let's look at an example smart contract which implements a simple
+naming system with just two kinds of transactions: _preorder_
+and _register_. The requirements of the system are as follows:
+
+1. Names may only be owned by one principal
+2. A register is only allowed if there is a corresponding preorder
+   with a matching hash
+3. A register transaction must be signed by the same principal who
+   paid for the preorder
+4. A preorder must have paid at least the price of the name. Names
+   are represented as integers, and any name less than 100000 costs
+   1000 microstacks, while all other names cost 100 microstacks.
+
+In this simple scheme, names are represented by integers, but in
+practice, a buffer would probably be used.
+
+```scheme
+(define burn-address '1111111111111111111114oLvT2)
+(define (price-function name)
+  (if (< name 1e5) 1000 100))
+
+(defmap name-map 
+  ((name integer)) ((buyer principal)))
+(defmap preorder-map
+  ((name-hash (buffer 160)) (buyer principal))
+  ((paid integer)))
+
+(define (tx-preorder 
+           (buyer-principal principal)
+           (name-hash (buffer 20))
+           (name-price integer))
+  (if (stacks-transfer!
+        buyer-principal name-price burn-address name-price)
+      (insert-entry! preorder-map
+        (tuple #name-hash name-hash
+               #buyer buyer-principal)
+        (tuple #paid name-price))
+      false))
+
+(define (tx-register 
+           (buyer-principal principal)
+           (recipient-principal principal)
+           (name integer)
+           (salt integer)
+  (if (signed-by? buyer-principal)
+    (let ((preorder-lookup
+            (tuple #name-hash (hash160 name salt) 
+                   #buyer      buyer-principal))
+          (preorder-entry 
+            (fetch-entry preorder-map preorder-lookup))
+          (name-entry 
+            (fetch-entry name-map (tuple #name name))))
+      (if (and
+           ;; must be preordered
+           (not (eq? preorder-entry) 'null)
+           ;; name shouldn't *already* exist
+           (eq? name-entry 'null)
+           ;; preorder must have paid enough
+           (<= (price-funcion name) 
+               (get #paid preorder-entry)))
+          (begin
+            (delete-entry! preorder-lookup)
+            (insert-entry! name-table
+              (tuple #name name)
+              (tuple #owner recipient)))
+          false))
+    false)))
+```
```

Comment:
We should probably also say here that Blockstack PBC intends to supply a full BNS smart contract, as well as a formal proof that certain desirable properties hold (e.g. "names are globally unique", "a revoked name cannot be updated or transferred", "names cost stacks based on their namespace price function", "only the principal can reveal a name on registration", etc.).

Reply:
Yep, will add